feature-governance-impl/src/main/java/io/novafoundation/nova/feature_governance_impl/presentation/tindergov/basket/TinderGovBasketViewModel.kt

```diff
@@ -0,0 +1,158 @@
+        if (it) {
+            resourceManager.getString(R.string.common_done)
+        } else {
+            resourceManager.getString(R.string.common_edit)
+        }
+    }
+
+    val basketFlow = combine(availableToVoteReferendaFlow, basketItemsFlow) { referendaById, basketItems ->
+        basketItems.mapNotNull {
+            val referendum = referendaById[it.referendumId] ?: return@mapNotNull null
+            mapBasketItem(governanceSharedState.chainAsset(), it, referendum)
+        }
+    }
+
+    init {
+        validateReferendaAndRemove()
+    }
+
+    fun backClicked() {
+        router.back()
+    }
+
+    fun onItemClicked(item: TinderGovBasketRvItem) {
+        showMessage("Not implemented yet")
+    }
+
+    fun onItemDeleteClicked(item: TinderGovBasketRvItem) {
+        launch {
+            val refId = referendumFormatter.formatId(item.id)
+            val title = resourceManager.getString(R.string.tinder_gov_basket_remove_item_confirm_title, refId)
+
+            if (removeReferendumAction.awaitAction(title)) {
+                val referendum = basketItemsFlow.first()
+                    .associateBy { it.referendumId }
+                    .get(item.id) ?: return@launch
+
+                interactor.removeReferendumFromBasket(referendum)
+
+                closeScreenIfBasketIsEmpty()
+            }
+        }
+    }
+
+    fun voteClicked() {
+        showMessage("Not implemented yet")
+    }
+
+    fun toggleEditMode() {
+        inEditModeFlow.toggle()
+    }
+
+    private fun mapBasketItem(
+        chainAsset: Chain.Asset,
+        item: TinderGovBasketItem,
+        referendum: ReferendumPreview
+    ): TinderGovBasketRvItem {
+        val voteType = when (item.voteType) {
+            VoteType.AYE -> resourceManager.getString(R.string.tinder_gov_aye_format).withColor(R.color.text_positive)
+            VoteType.NAY -> resourceManager.getString(R.string.tinder_gov_nay_format).withColor(R.color.text_negative)
+            VoteType.ABSTAIN -> resourceManager.getString(R.string.tinder_gov_abstain_format).withColor(R.color.text_secondary)
+        }
+
+        val votesAmount = chainAsset.amountFromPlanks(item.amount)
+        val votes = votersFormatter.formatVotes(votesAmount, item.voteType, item.conviction)
+            .withColor(R.color.text_secondary)
+
+        return TinderGovBasketRvItem(
+            id = item.referendumId,
+            idStr = referendumFormatter.formatId(item.referendumId),
+            title = referendumFormatter.formatReferendumName(referendum),
+            subtitle = SpannableFormatter.format(voteType, votes)
+        )
+    }
+
+    private fun String.withColor(@ColorRes color: Int): CharSequence {
+        return toSpannable(colorSpan(resourceManager.getColor(color)))
+    }
+
+    private fun validateReferendaAndRemove() {
+        launch {
+            val basket = basketItemsFlow.first()
+                .associateBy { it.referendumId }
```

Comment:
You are using `associateBy` quite a lot recently : )

Here it doesnt make much sense as well since we're only using ` basket.filterKeys` which is still linear complexity